Prefer dynamic allocations over large STATIC arrays

vikramdattu opened the pull request:
## What changed
Replace large static arrays with dynamic allocations in signaling API call functions, gated behind the `-DPREFER_DYNAMIC_ALLOCS=ON` CMake option (default OFF).

## Why
Functions like `describeChannelLws`, `createChannelLws`, `getChannelEndpointLws`, `getIceConfigLws`, `deleteChannelLws`, `connectSignalingChannel`, `joinStorageSessionLws`, and `describeMediaStorageConfLws` each allocate large stack arrays (`CHAR url[MAX_URI_CHAR_LEN + 1]` and `CHAR paramsJson[MAX_JSON_PARAMETER_STRING_LEN]`) that consume significant stack space. On memory-constrained platforms (ESP32), this is a major issue.

## How
- **LwsApiCalls.c**: In each signaling API function, replaced `CHAR url[...]` and `CHAR paramsJson[...]` stack arrays with dynamically allocated buffers sized to exact need using `SNPRINTF(NULL, 0, ...)`. All buffers are freed in `CleanUp` paths.
- **Include.h**: Added `USE_DYNAMIC_URL` macro (enabled when `PREFER_DYNAMIC_ALLOCS` is set). Also includes `DYNAMIC_SIGNALING_PAYLOAD` for dynamic signaling message payload (also in PR #2140).
- **samples/common/Common.c**: Added payload alloc/free guards in `respondWithAnswer`, `onIceCandidateHandler`, `submitPendingIceCandidate`, and `signalingMessageReceived`.
- **CMakeLists.txt**: Wired up `PREFER_DYNAMIC_ALLOCS` CMake option.

## Testing done
- Builds with both `-DPREFER_DYNAMIC_ALLOCS=ON` and default (OFF)
- Clang-format compliance check passes
- Rebased onto latest `develop` branch

## Related PRs
- PR #2140 — subset of this PR, covers only `DYNAMIC_SIGNALING_PAYLOAD`

vikramdattu commented:
@sirknightj @unicornss would love to get your review on this particular PR.

sirknightj commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -2279,7 +2491,9 @@ STATUS receiveLwsMessage(PSignalingClient pSignalingClient, PCHAR pMessage, UINT
-
+#ifdef DYNAMIC_SIGNALING_PAYLOAD
+        SAFE_MEMFREE(pSignalingMessageWrapper->receivedSignalingMessage.signalingMessage.payload);
+#endif
```
In the sample application, would you be able to demo how this one works for others with the ifdef, maybe add something in the readme about the new option.

sirknightj replied:
Thank you, I will give it a try

sirknightj replied:
I ran the sample and it works fine.

The test is failing for me oddly enough:

```
./tst/webrtc_client_test --gtest_filter="SignalingApiFunctionalityTest.SignalingMessageParsing_ParseOfferSuccess"
Note: Google Test filter = SignalingApiFunctionalityTest.SignalingMessageParsing_ParseOfferSuccess
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from SignalingApiFunctionalityTest
[ RUN      ] SignalingApiFunctionalityTest.SignalingMessageParsing_ParseOfferSuccess
Address of receivedMessage: 0x16b3c5f18
Address of payload member: 0x16b3c6128
Allocation address: 0x600000eac0e0
Allocated payload points to: 0x600000eac0e0
About to free pointer: 0x600000eac0e0
webrtc_client_test(98158,0x20bebcc80) malloc: *** error for object 0x600000eac0d8: pointer being freed was not allocated
webrtc_client_test(98158,0x20bebcc80) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    98158 abort      ./tst/webrtc_client_test 
```

It seems like 8 bytes went missing.

```
(lldb) frame select 6
frame #6: 0x00000001008838b4 libkvsWebrtcClient.dylib`defaultMemFree(ptr=0x0000600000f20048) at Allocators.c:36:5
(lldb) p ptr
(void *) 0x0000600000f20048
(lldb) frame select 8
frame #8: 0x000000010014aa90 webrtc_client_test`com::amazonaws::kinesis::video::webrtcclient::SignalingApiFunctionalityTest_SignalingMessageParsing_ParseOfferSuccess_Test::TestBody(this=0x000000012b809400) at SignalingApiFunctionalityTest.cpp:3657:5
   3654     EXPECT_EQ(SIGNALING_MESSAGE_TYPE_OFFER, receivedMessage.signalingMessage.messageType);
   3655 #ifdef DYNAMIC_SIGNALING_PAYLOAD
   3656     printf("About to free pointer: %p\n", receivedMessage.signalingMessage.payload);
-> 3657     SAFE_MEMFREE(receivedMessage.signalingMessage.payload);
   3658 #endif
   3659 }
   3660 
(lldb) p receivedMessage.signalingMessage.payload
error: <user expression 5>:1:17: no member named 'signalingMessage' in '(unnamed struct)'
    1 | receivedMessage.signalingMessage.payload
      | ~~~~~~~~~~~~~~~ ^
(lldb) x/4gx &receivedMessage.signalingMessage.payload
0x16fdfdea8: 0x0000600000f20050 0x0000000000000000
0x16fdfdeb8: 0x0000000000000000 0x0000000000000000
(lldb) memory read --size 8 --format x --count 4 receivedMessage.signalingMessage.payload
0x600000f20050: 0x0000006f6c6c6548 0x0000000000000000
0x600000f20060: 0x0000000000000000 0x0000000000000000
(lldb) 
```

I was able to get around it by replacing the `SAFE_MEMFREE` in that test with a direct `free()`. `MEMFREE` seems to also run into the segfault. Though, that doesn't seem ideal.

sirknightj replied:
Did a bit more digging, though I have not found a root cause yet.

Tested on M1 Pro Macbook -- built with: `cmake .. -DUSE_OPENSSL=OFF -DUSE_MBEDTLS=ON -DCMAKE_BUILD_TYPE=Debug -DBUILD_TEST=ON -DENABLE_AWS_SDK_IN_TESTS=OFF -DPREFER_DYNAMIC_ALLOCS=ON`

Running with address sanitizer:
```
./tst/webrtc_client_test --gtest_filter="SignalingApiFunctionalityTest.SignalingMessageParsing_ParseOfferSuccess"
webrtc_client_test(84998,0x20bebcc80) malloc: nano zone abandoned due to inability to reserve vm space.
Note: Google Test filter = SignalingApiFunctionalityTest.SignalingMessageParsing_ParseOfferSuccess
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from SignalingApiFunctionalityTest
[ RUN      ] SignalingApiFunctionalityTest.SignalingMessageParsing_ParseOfferSuccess
=================================================================
==84998==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x6020000034c8 in thread T0
    #0 0x1061b8d40 in free+0x98 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x54d40)
    #1 0x105bc5594 in defaultMemFree Allocators.c:36
    #2 0x105bc9f58 in instrumentedAllocatorsMemFree InstrumentedAllocators.c:187
    #3 0x104f281f8 in com::amazonaws::kinesis::video::webrtcclient::SignalingApiFunctionalityTest_SignalingMessageParsing_ParseOfferSuccess_Test::TestBody() SignalingApiFunctionalityTest.cpp:3650
    #4 0x10505281c in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x90 (webrtc_client_test:arm64+0x10051681c)
    #5 0x10501c9c0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x5c (webrtc_client_test:arm64+0x1004e09c0)
    #6 0x10501c910 in testing::Test::Run()+0xbc (webrtc_client_test:arm64+0x1004e0910)
    #7 0x10501d860 in testing::TestInfo::Run()+0x12c (webrtc_client_test:arm64+0x1004e1860)
    #8 0x10501e9c8 in testing::TestSuite::Run()+0x144 (webrtc_client_test:arm64+0x1004e29c8)
    #9 0x10502c478 in testing::internal::UnitTestImpl::RunAllTests()+0x420 (webrtc_client_test:arm64+0x1004f0478)
    #10 0x1050593a0 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x90 (webrtc_client_test:arm64+0x10051d3a0)
    #11 0x10502be18 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x5c (webrtc_client_test:arm64+0x1004efe18)
    #12 0x10502bd04 in testing::UnitTest::Run()+0xd4 (webrtc_client_test:arm64+0x1004efd04)
    #13 0x104ffb350 in RUN_ALL_TESTS() gtest.h:2293
    #14 0x104ffb1c8 in main main.cpp:88
    #15 0x19cc4ab48  (<unknown module>)

0x6020000034c8 is located 8 bytes before 7-byte region [0x6020000034d0,0x6020000034d7)
allocated by thread T0 here:
    #0 0x1061b8fd0 in calloc+0x9c (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x54fd0)
    #1 0x10565ab70 in defaultMemCalloc+0x1c (libkvsCommonLws.1.5.4.dylib:arm64+0xeb70)
    #2 0x1056d9370 in parseSignalingMessage LwsApiCalls.c:2207
    #3 0x104f276b8 in com::amazonaws::kinesis::video::webrtcclient::SignalingApiFunctionalityTest_SignalingMessageParsing_ParseOfferSuccess_Test::TestBody() SignalingApiFunctionalityTest.cpp:3642
    #4 0x10505281c in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x90 (webrtc_client_test:arm64+0x10051681c)
    #5 0x10501c9c0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x5c (webrtc_client_test:arm64+0x1004e09c0)
    #6 0x10501c910 in testing::Test::Run()+0xbc (webrtc_client_test:arm64+0x1004e0910)
    #7 0x10501d860 in testing::TestInfo::Run()+0x12c (webrtc_client_test:arm64+0x1004e1860)
    #8 0x10501e9c8 in testing::TestSuite::Run()+0x144 (webrtc_client_test:arm64+0x1004e29c8)
    #9 0x10502c478 in testing::internal::UnitTestImpl::RunAllTests()+0x420 (webrtc_client_test:arm64+0x1004f0478)
    #10 0x1050593a0 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x90 (webrtc_client_test:arm64+0x10051d3a0)
    #11 0x10502be18 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x5c (webrtc_client_test:arm64+0x1004efe18)
    #12 0x10502bd04 in testing::UnitTest::Run()+0xd4 (webrtc_client_test:arm64+0x1004efd04)
    #13 0x104ffb350 in RUN_ALL_TESTS() gtest.h:2293
    #14 0x104ffb1c8 in main main.cpp:88
    #15 0x19cc4ab48  (<unknown module>)

SUMMARY: AddressSanitizer: bad-free (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x54d40) in free+0x98
==84998==ABORTING
[1]    84998 abort      ./tst/webrtc_client_test 
```

Replacing the `SAFE_MEMFREE` with:
* MEMFREE -- ❌ 
* globalMemFree -- ❌ 
* defaultMemFree -- ✅ 
* free -- ✅ 

Strange that it only shows up in the tests and not in the sample..

sirknightj replied:
I added this immediately above the test (SignalingApiFunctionalityTest:3631) and it seems to fix the issue:

```
VOID safeMemFree(VOID* ptr) {
    defaultMemFree(ptr);
}

memFree globalMemFree = safeMemFree;
```

So it would be logical to conclude that the issue might be related to the type definitions or maybe an <include\> ordering issue?

In PIC, PVOID and VOID* are declared the same -
https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/c98c2a256a7bb3dfc4db41ae26d45e28ac7eec56/src/common/include/com/amazonaws/kinesis/video/common/CommonDefs.h#L183

vikramdattu replied:
@sirknightj Eureka!! 
I found out the reason! Went ahead debugging with lldb and found out:

The tests use `instrumentedAllocator` versions of allocators whereas the SDK is using global Allocators.
This free here is decrementing the pointer before free: https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/c98c2a256a7bb3dfc4db41ae26d45e28ac7eec56/src/utils/src/InstrumentedAllocators.c#L181

sirknightj replied:
Interesting finding!

I do see that in the test setup, it call `setInstrumentedAllocators()`, so the `MEMCALLOC` should have been set to point to the allocated one to match..

https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/cf817bc5d18f3e4bd499c6b0f9a68c6f4d7e01de/tst/WebRTCClientTestFixture.cpp#L57

And `initKvsWebrtc()` also calls it: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/cf817bc5d18f3e4bd499c6b0f9a68c6f4d7e01de/tst/WebRTCClientTestFixture.cpp#L68 https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/cf817bc5d18f3e4bd499c6b0f9a68c6f4d7e01de/src/source/PeerConnection/PeerConnection.c#L1773

sirknightj replied:
Hi @vikramdattu, it seems the change for the CI was accidentally removed in the latest changes, could you re-add it please? Want to make sure we have automated testing (avoid manual testing) for all features if possible, thanks! 
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/2146#discussion_r2155842250

vikramdattu replied:
Added, can you give it a run please?

sirknightj reviewed (commented):
Good memory optimizations enhancements that reduce the allocated memory per URL from:

```
#define MAX_URI_CHAR_LEN 10000
#define MAX_JSON_PARAMETER_STRING_LEN (10 * 1024)
```

about 20k

Down to whatever is required based on the calculated length

MushMal commented:
I believe this is a good change. Previously, the codebase was optimized to avoid dynamic allocations due to the cost impact on many small footprint platforms in favor of the predictable memory footprint. Most platforms do provide a dynamic memory allocators and if a particular platform still has eprom type of memory only, the allocators for the SDK can be overridden to utilize the sub-allocator the SDK provides (AIV heap implementation). There are many samples of this in the tests that validate the memory leakage.

vikramdattu commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -763,12 +769,28 @@ STATUS describeChannelLws(PSignalingClient pSignalingClient, UINT64 time)
 
+#if PREFER_DYNAMIC_ALLOCS
+    // allocate memory for url
+    UINT32 urlLen = STRLEN(pSignalingClient->pChannelInfo->pControlPlaneUrl) + STRLEN(DESCRIBE_SIGNALING_CHANNEL_API_POSTFIX);
```
@sirknightj  Right, for fix strings, the compiler does take care of the optimisation. So, should be okay to have STRLEN.
It looks hack-ish/magic number to have hard-coded numbers.

vikramdattu commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -2061,6 +2269,10 @@ STATUS receiveLwsMessage(PSignalingClient pSignalingClient, PCHAR pMessage, UINT
 
     CHK(NULL != (pSignalingMessageWrapper = (PSignalingMessageWrapper) MEMCALLOC(1, SIZEOF(SignalingMessageWrapper))), STATUS_NOT_ENOUGH_MEMORY);
 
+#ifdef DYNAMIC_SIGNALING_PAYLOAD
```
If you like me to break the commit into two, this PR contains changes only related to dynamic URLs, and move the DYNAMIC_PAYLOAD change to other PR, (this will allow changes without needing any change in Tests/Applications/CMake).
Your thoughts.

vikramdattu committed:
Prefer dynamic allocations over large static arrays

Replace large stack-allocated arrays with dynamic allocations in signaling
API functions, gated behind -DPREFER_DYNAMIC_ALLOCS=ON (default OFF).

When enabled:
- DYNAMIC_SIGNALING_PAYLOAD: SignalingMessage.payload becomes a dynamically
  allocated PCHAR instead of CHAR[MAX_SIGNALING_MESSAGE_LEN + 1]
- USE_DYNAMIC_URL: url and paramsJson buffers in LwsApiCalls are allocated
  to exact size needed via SNPRINTF(NULL, 0, ...) instead of using
  MAX_URI_CHAR_LEN and MAX_JSON_PARAMETER_STRING_LEN stack arrays

This significantly reduces stack and peak memory usage on constrained
platforms (e.g. ESP32).

To keep ownership of the payload buffer on the SDK side, this commit also
introduces a public helper:

    PUBLIC_API VOID freeSignalingMessagePayload(PSignalingMessage pMessage);

Applications must call this helper to release the payload that the SDK
allocated when DYNAMIC_SIGNALING_PAYLOAD is enabled. The declaration is
unconditional, so callers do not need to wrap the call in an #ifdef; in
the static-buffer build the helper is a no-op. This avoids cross-allocator
mismatches when the SDK and the application link different allocator
implementations (for example a test binary that wraps malloc/free).

USE_DYNAMIC_URL remains transparent to applications.

vikramdattu committed:
ci: add PREFER_DYNAMIC_ALLOCS coverage to CI and samples

Add a dedicated CI job to run unit tests with dynamic allocations
enabled, and enable the flag in the valgrind sample check.

sirknightj commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -1017,18 +1071,31 @@ STATUS getChannelEndpointLws(PSignalingClient pSignalingClient, UINT64 time)
+    CHK(url != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    // Allocate memory for paramsJson
+    paramsJsonLen = SNPRINTF(NULL, 0, GET_CHANNEL_ENDPOINT_PARAM_JSON_TEMPLATE, pSignalingClient->channelDescription.channelArn,
```
paramsJson here is sized using `SIGNALING_CHANNEL_PROTOCOL` (14 chars), but when `storageStatus == TRUE` (line 1098), it's written with `SIGNALING_CHANNEL_PROTOCOL_W_MEDIA_STORAGE` (24 chars), which would overflow the buffer by 10 bytes.

sirknightj commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -763,12 +764,24 @@ STATUS describeChannelLws(PSignalingClient pSignalingClient, UINT64 time)
 
     CHK(pSignalingClient != NULL, STATUS_NULL_ARG);
 
+    // Allocate memory for url
```
Maybe forgot to add the `USE_DYNAMIC_URL` macro?

vikramdattu replied:
`USE_DYNAMIC_URL` was a leftover define from before the static URL path was removed — never actually gated anything. Dropped in this push.

sirknightj commented on samples/common/Common.c:
```diff
@@ -1690,6 +1717,9 @@ STATUS signalingMessageReceived(UINT64 customData, PReceivedSignalingMessage pRe
+    if (pReceivedSignalingMessage != NULL) {
+        freeSignalingMessagePayload(&pReceivedSignalingMessage->signalingMessage);
+    }
     SAFE_MEMFREE(pReceivedSignalingMessageCopy);
```
If an error occurs after line 1681 allocates pReceivedSignalingMessageCopy->signalingMessage.payload but before pReceivedSignalingMessageCopy is set to NULL (line 1692),  this SAFE_MEMFREE frees the struct but leaks its payload.

Suggest to also add a check to free the payload:
```c
if (pReceivedSignalingMessageCopy != NULL) {
    freeSignalingMessagePayload(&pReceivedSignalingMessageCopy->signalingMessage);
}
```

sirknightj commented on samples/common/Common.c:
```diff
@@ -343,13 +351,21 @@ VOID onIceCandidateHandler(UINT64 customData, PCHAR candidateJson)
         STRNCPY(message.peerClientId, pSampleStreamingSession->peerId, MAX_SIGNALING_CLIENT_ID_LEN);
         message.payloadLen = (UINT32) STRNLEN(candidateJson, MAX_SIGNALING_MESSAGE_LEN);
+#ifdef DYNAMIC_SIGNALING_PAYLOAD
+        message.payload = (PCHAR) MEMALLOC(message.payloadLen + 1);
```
Concern if `candidateJson` is exactly `payloadLen` chars

Suggestion: Either use `MEMCALLOC(1, message.payloadLen + 1)` or add `message.payload[message.payloadLen] = '\0';` after the `STRNCPY` on 358 to guarantee null termination of the payload.

sirknightj reviewed (commented):
Wondering if you had the chance to measure the memory savings with the flag on/off?

sirknightj commented:
Looks like some of the tests are failing, please take a look!

sirknightj commented on src/include/com/amazonaws/kinesis/video/webrtcclient/Include.h:
```diff
@@ -2371,6 +2367,28 @@ PUBLIC_API STATUS freeSignalingClient(PSIGNALING_CLIENT_HANDLE);
+ *
+ * Safe to call with a NULL pMessage.
+ *
+ * @param[in,out] pMessage SignalingMessage whose payload should be released.
```
Consider adding a note to the API doc that callers must invoke freeSignalingMessagePayload() regardless of whether parseSignalingMessage() succeeded or failed, since the payload may have been allocated before a later parse step failed. The internal comment in LwsApiCalls.c (line 2277) documents this intent, but external SDK consumers won't see that.

vikramdattu replied:
Done in 2cda6dc — surfaced the failure-path contract in the public doc on Include.h. Thanks for catching this.

sirknightj reviewed (commented):
Looks like some of the CI is still failing

vikramdattu committed:
style: apply clang-format to two SNPRINTF call sites

The two SNPRINTF call sites in getChannelEndpointLws() exceeded the
project's line-length wrap point in the dynamic-allocs commit. Apply
clang-format -i to satisfy the clang-format-check CI job.

vikramdattu commented:
Pushed an update:
- Rebased on `develop` (picked up #2292 and #2279).
- Fixed `clang-format-check`: two `SNPRINTF` sites in `getChannelEndpointLws()` were too long; ran `clang-format -i`.

On memory savings: per LWS API call, the `url` + `paramsJson` stack arrays drop from ~20 KB combined (10001 + 10240 bytes) to exact-size heap allocations of a few hundred bytes each. Per `SignalingMessage`, the inline payload buffer goes from `18751` bytes to an 8-byte pointer + an exact-size alloc — so ~14–18 KB saved per message for typical SDP/ICE payloads. On our ESP32-P4 builds that's the difference between barely fitting and having real headroom.

On the remaining CI red: I ran the full suite locally on macOS with real AWS creds — 374/377 pass. The 3 that fail are `*AuthExpiration` variants, and they fail identically on `develop` with `PREFER_DYNAMIC_ALLOCS=OFF`, so they're pre-existing (the tests expect single-shot signaling but the SDK retries on auth expiry — needs mocked auth). The other 35 failing in `dynamic-allocs-check` CI all bail at exactly 401 ms with `mAccessKeyIdSet=0` — looks like the job isn't getting AWS creds. `valgrind-check` smells the same (master exits before SIGINT).

vikramdattu committed:
docs: clarify freeSignalingMessagePayload contract on failure paths

Per #2146 review: external consumers couldn't see the LwsApiCalls.c
comment explaining that the parser zeros the struct up front so the
helper is safe to invoke after any return path. Surface that contract
in the public header doc so callers know they must invoke
freeSignalingMessagePayload() even when the receive/parse step
returned a non-SUCCESS status.

sirknightj commented on .github/workflows/ci.yml:
```diff
@@ -466,6 +466,35 @@ jobs:
+      - name: Setup cmake
+        uses: jwlawson/actions-setup-cmake@v2
+        with:
+          cmake-version: '3.x'
```
We'll need the configure AWS credentials step here. You can copy it from the others.

sirknightj reviewed (commented):
Thanks for the details. There are 2 items that need to be addressed:

1. Dynamic allocs check: CI is missing the configure-aws-credentials step so it doesn't have the permissions to run
2. The viewer sample + valgrind seems to be failing. I enabled verbose logs in a run here. Can you check: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/actions/runs/26302497447/job/77430927183

> 2026-05-22 17:33:20.951 PROFILE main(): [Initialize signaling client and connect to the signaling channel] Time taken: 2228.94 ms
2026-05-22 17:33:20.951 INFO    main(): [KVS Master] Channel demo-channel-gh-actions set up done 
2026-05-22 17:33:20.952 VERBOSE signalingClientGetCurrentState(): Signaling Client Get Current State
2026-05-22 17:33:20.952 VERBOSE signalingClientGetCurrentState(): Current state: 0x0000000000000100
2026-05-22 17:33:21.456 PROFILE createRtcCertificate(): [Certificate creation time] Time taken: 38.13 ms
2026-05-22 17:33:21.457 VERBOSE pregenerateCertTimerCallback(): New certificate has been pre-generated and added to the queue
2026-05-22 17:33:25.954 VERBOSE signalingClientGetCurrentState(): Signaling Client Get Current State
2026-05-22 17:33:25.954 VERBOSE signalingClientGetCurrentState(): Current state: 0x0000000000000100
2026-05-22 17:33:30.949 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 10
/home/runner/work/_temp/a194e02a-cf15-4ff5-a921-895ea5c17f38.sh: line 8: 11898 Segmentation fault      (core dumped) valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose --log-file=valgrind-viewer.txt ./samples/kvsWebrtcClientViewer demo-channel-gh-actions

sirknightj commented:
Looks like some tests are failing in the dynamic-allocs test:
```c
Tearing down test: SignalingApiFunctionalityTest

2026-05-26 18:44:51.211 WARN    resetInstrumentedAllocators(): Possible memory leak of size 13
2026-05-26 18:44:51.211 ERROR   resetInstrumentedAllocators(): operation returned status code: 0x0000001e
/home/runner/work/amazon-kinesis-video-streams-webrtc-sdk-c/amazon-kinesis-video-streams-webrtc-sdk-c/tst/WebRTCClientTestFixture.cpp:119: Failure
Expected equality of these values:
  ((UINT32) 0x00000000)
    Which is: 0
  resetInstrumentedAllocators()
    Which is: 30
[  FAILED  ] SignalingApiFunctionalityTest.goAwayEmulation (2570 ms)
[----------] 2 tests from SignalingApiFunctionalityTest (4834 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (4834 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] SignalingApiFunctionalityTest.iceReconnectEmulation
[  FAILED  ] SignalingApiFunctionalityTest.goAwayEmulation
```

vikramdattu committed:
IceAgent: heap-allocate turnChannelData scratch buffer

incomingRelayedDataHandler declared TurnChannelData turnChannelData[512]
on the stack (~16 KB: 512 x 32-byte entries). That single local overflows
constrained worker-thread stacks (e.g. the ICE timer-queue executor that
runs this handler) before any call depth.

Move it to the heap (MEMCALLOC + SAFE_MEMFREE), keeping the same worst-case
capacity. Typical usage is ~4 entries per the existing comment; this only
changes where the buffer lives, not behavior.
